source/state_representation/src/trajectory/CartesianTrajectory.cpp

```diff
@@ -0,0 +1,210 @@
+#include "state_representation/trajectory/CartesianTrajectory.hpp"
+#include "state_representation/space/cartesian/CartesianState.hpp"
+
+#include "state_representation/exceptions/EmptyStateException.hpp"
+#include "state_representation/exceptions/IncompatibleReferenceFramesException.hpp"
+
+namespace state_representation {
+CartesianTrajectory::CartesianTrajectory() {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+}
+
+CartesianTrajectory::CartesianTrajectory(const std::string& name, const std::string& reference_frame)
+    : TrajectoryBase<CartesianTrajectoryPoint>(name), reference_frame_(reference_frame) {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+}
+
+CartesianTrajectory::CartesianTrajectory(
+    const std::string& name, const CartesianState& point, const std::chrono::nanoseconds& duration
+)
+    : TrajectoryBase<CartesianTrajectoryPoint>(name) {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+  if (point.is_empty()) {
+    throw exceptions::EmptyStateException("The Cartesian state provided is empty");
+  }
+  this->reference_frame_ = point.get_reference_frame();
+  this->add_point(point, duration);
+}
+
+CartesianTrajectory::CartesianTrajectory(
+    const std::string& name, const std::vector<CartesianState>& points,
+    const std::vector<std::chrono::nanoseconds>& durations
+)
+    : TrajectoryBase<CartesianTrajectoryPoint>(name) {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+  if (points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (std::ranges::any_of(points, [&](const auto& p) { return p.is_empty(); })) {
+    throw exceptions::EmptyStateException("Vector contains at least one point that is empty");
+  } else if (!std::ranges::all_of(points, [&](const auto& p) {
+               return p.get_reference_frame() == points.front().get_reference_frame();
+             })) {
+    throw exceptions::IncompatibleReferenceFramesException("Incompatible reference frames within the points vector");
+  }
+  this->reference_frame_ = points[0].get_reference_frame();
+  try {
+    this->add_points(points, durations);
+  } catch (...) {
+    throw;
+  }
+}
+
+const std::string& CartesianTrajectory::get_reference_frame() const {
+  return this->reference_frame_;
+}
+
+void CartesianTrajectory::set_reference_frame(const CartesianPose& pose) {
+  if (pose.is_empty()) {
+    throw exceptions::EmptyStateException("Pose is empty");
+  } else if (this->get_size() == 0) {
+    throw exceptions::EmptyStateException("Trajectory is empty");
+  }
+  this->reference_frame_ = pose.get_reference_frame();
+  auto points = this->get_points();
+  for (auto& point : points) {
```

Comment:
Question: Don't we have a non const getter/operator[] to do that in place? Otherwise I don't understand why we needed to define it in the first place

Reply:
Not for state types, no. We only have a reference on the flat data operator[], the operator[] overloads in derived classes have to create a `std::vector<StateT>` at which point the reference is not something we can use